### PR TITLE
Add Tinkerer Log download operation

### LINKED ISSUE #814: Write tinkerer command to fetch logs
**Description:**
We need a tinkerer command to fetch carbon logs, mvn build logs, etc. from deployment instances of the stack to TestGrid.

**Suggested Labels:**
New Feature, Priority High

**Suggested Assignees:**
SameeraW

**Affected Product Version:**
m31
**OS, DB, other environment details and versions:**    
N/A
**Steps to reproduce:**
N/A

**Related Issues:**
#817

### PR DESCRIPTION
**Pull request by sameerawickramasekara**
**Purpose**

Resolves #814 

**Goals**
This changes will enable downloading log files from the instances to the TestGrid environment.
All the files in the predefined log location for each test type will be downloaded.

**Approach**

The current implementation will use ssh access to the instance to download the logs. it is required the tinkerer Webapp to be hosted in the same environment as test grid for this approach to be successful.
A new testgrid.yaml property is introduced to indicate the location of the security key file (.pem file).
This property needs to be added to the job-config.properties file in the workspace

The AMIs also need to to have the agent pre-installed and should contain the
AGENT_READY value set.
Furthermore, the username of the instance needs to be also added as a TAG value.

Ex :
```
 Ubuntu ubuntu
CentOS centos
```

**Security checks**
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java [hunk 1]
```diff
@@ -139,7 +193,7 @@ public void execute(TestPlan testPlan, InfrastructureConfig infrastructureConfig
         persistTestPlanStatus(testPlan);
 
         //cleanup
-        releaseInfrastructure(testPlan, infrastructureProvisionResult, deploymentCreationResult);
+//        releaseInfrastructure(testPlan, infrastructureProvisionResult, deploymentCreationResult);
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Commented line of code

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java [hunk 2]
```diff
@@ -618,7 +672,40 @@ private static String getHumanReadableTimeDiff(long timeDifferenceMilliseconds)
...
             return diffDays + " days" + " " + (diffHours % 24) + " hours";
         }
+    }
+
+    /**
+     *
+     * @param infraParameters
+     * @return
+     */
+    private OSCategory getOsCatagory(String infraParameters) {
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Would it be better if we use 'getOSCategory' ? :)

File: core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java [hunk 3]
```diff
@@ -618,7 +672,40 @@ private static String getHumanReadableTimeDiff(long timeDifferenceMilliseconds)
...
+        if (infraParameters.contains(OSCategory.WINDOWS.toString().toLowerCase(Locale.ENGLISH))) {
+            return OSCategory.WINDOWS;
+        } else {
+            return OSCategory.UNIX;
+        }
+    }
+
+    /**
+     *
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/core/TestPlanExecutor.java]: Empty Javadoc lines

File: core/src/main/java/org/wso2/testgrid/tinkerer/TinkererClient.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.testgrid.tinkerer;
+
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/tinkerer/TinkererClient.java]: We can reduce one empty line. :)

File: core/src/main/java/org/wso2/testgrid/tinkerer/UnixClient.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                String operationPath = this.getTinkererBase() + "test-plan/" + agent.getTestPlanId()
+                        + "/agent/" + agent.getInstanceName() + "/operation";
+                if (TestGridConstants.TEST_TYPE_FUNCTIONAL.equals(testType)) {
+                    //get the carbon.home from the instance where the product is running
+                    Response execute = Request.Post(operationPath).setHeader
+                            (HttpHeaders.CONTENT_TYPE, "application/json")
+                            .setHeader(HttpHeaders.AUTHORIZATION, "Basic " + Base64.getEncoder()
+                                    .encodeToString(StringUtil.concatStrings(this.getTinkererUserName(),
+                                            ":", this.getTinkererPassword()).getBytes(Charset.defaultCharset())))
+                            .bodyString("{\"code\":\"SHELL\",\"request\":\"ps -ef |grep 'carbon.home'\"}",
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/tinkerer/UnixClient.java]: Would it be better if we create a JSON object and add those elements and then convert it back to String? :)

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/api/DeploymentTinkerer.java [hunk 6]
```diff
@@ -150,4 +154,59 @@ public Response sendOperation(@PathParam("testPlanId") String testPlanId,
...
+                if (bastianIP != null) {
+                    //if there is a bastian entry add the host entry
+                    SSHHelper.addConfigEntry(keyFile, bastianIP, agent);
+                } else {
+                    //if there is no bastian entry then it does not need a proxy entry
+                    SSHHelper.addConfigEntry(keyFile, agent);
+                }
+                ShellExecutor executor = new ShellExecutor();
+                executor.executeCommand("ssh host:" + agent.getInstanceId() + " \"cat " + source
+                        + "\" &> " + destination + "");
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/api/DeploymentTinkerer.java]: This assumes a linux environment. What if that's not the case?

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java [hunk 7]
```diff
@@ -81,4 +84,50 @@ public String getInstanceName(String region, String instanceId) {
...
+            List<String> imageIds = new ArrayList<>();
+            imageIds.add(imageId.get());
+            //get ami from the instance ID
+            DescribeImagesRequest imageReq = new DescribeImagesRequest();
+            imageReq.setImageIds(imageIds);
+            //Get the Tag containing the username from the ami
+            DescribeImagesResult describeImagesResult = amazonEC2.describeImages(imageReq);
+            Optional<String> userName = describeImagesResult.getImages().stream()
+                    .flatMap(image -> image.getTags().stream())
+                    .filter(tag -> tag.getKey().equals("USERNAME"))
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java]: The tag name should be made a constant.

File: core/src/main/java/org/wso2/testgrid/tinkerer/UnixClient.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                                                        bastianIP + "\"" : "") + "}}"
+                                        , ContentType.APPLICATION_JSON)
+                                .execute();
+                        logger.debug("Download Location :" + destination);
+                    }
+                    logger.info("Successfully downloaded all log files ");
+                }
+            }
+        } catch (IOException e) {
+            throw new TinkererOperationException("Error occurred while executing the Tinkerer operation ", e);
```
Comment (pasindujw) [inline core/src/main/java/org/wso2/testgrid/tinkerer/UnixClient.java]: Shall we add some more details to the Error such as , operation name, error occurred in which agent from the deploymentCreationResult, log file name, etc.. :)

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+     * instance.
+     * @param keyContent the content of the security key file (.pem file) that is used to access the instance
+     * @param bastianIP The IP address of the bastian instance.
+     * @param agent The {@link Agent} object corresponding to the remote agent running on the instance
+     * @throws IOException When there is an error creating the config entry
+     */
+    public static void addConfigEntry(String keyContent, String bastianIP, Agent agent) throws IOException {
+
+        Path config = Paths.get(System.getProperty("user.home"), ".ssh", "config");
+        Path keyPath = Paths.get(System.getProperty("user.home"), ".ssh", agent.getInstanceId() + "-key.pem");
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java]: As a general rule, it's good if we do not touch system/user config files that may have its own lifecycle than the LC of the test-plan. The ~/.ssh/config is a user configuration file. 

IMO, we should create this file within the test-plan's workspace, and provide it as an input into SSH command. You can use _[-F](https://www.ssh.com/ssh/command/)_ cmd argument for that purpose. Shall we create a new ticket to track this?

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/SessionManager.java [hunk 10]
```diff
@@ -207,4 +208,30 @@ private static String getInstanceName(String provider, String region, String ins
...
+     * @param provider Provider Name
+     * @param region Region of the instance
+     * @param instanceId ID of the instance
+     * @return Username of the instance
+     */
+    private static String getInstanceUser(String provider, String region, String instanceId) {
+
+        Provider infraProvider = null;
+        switch (provider) {
+            case "aws":
```
Comment (pasindujw) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/SessionManager.java]: By any chance, can we receive "aws" with uppercase/ camel case here? If so, we may need to check with case.

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            //add Host entry
+            String entry = String.format("Host %s%n  StrictHostKeyChecking=no%n" +
+                            "  UserKnownHostsFile=/dev/null%n" +
+                            "  User %s%n" +
+                            "  HostName %s%n" +
+                            "  IdentityFile %s%n" +
+                            "  "
+                    , hostEntry, agent.getInstanceUser(), agent.getInstanceIp(),
+                    keyPath.toAbsolutePath().toString());
+            Files.write(config, entry.getBytes(StandardCharsets.UTF_8), StandardOpenOption.APPEND);
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java]: Seems the content is mostly duplicated.

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+     * @param entry Entry to be checked
+     * @return true if the entry is already existing , false otherwise
+     * @throws FileNotFoundException when there is an error finding the config file
+     */
+    private static boolean containsEntry(File file, String entry) throws FileNotFoundException {
+        Scanner scanner = new Scanner(file, StandardCharsets.UTF_8.name());
+        while (scanner.hasNextLine()) {
+            String line = scanner.nextLine();
+            if (line.contains(entry)) {
+                return true;
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java]: This is good. At the same time, we also have a Java 8 streaming API to do the same in a concise way. :)

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Saves the security key file in the local filesystem so it could be used when accessing the instances
+     *
+     * @param keyFile Contents of the key file
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java]: *base64 encoded

File: core/src/main/java/org/wso2/testgrid/tinkerer/TinkererClient.java (hunk 4 above)
Comment (sameerawickramasekara) [reply core/src/main/java/org/wso2/testgrid/tinkerer/TinkererClient.java]: Made the change and pushed to the branch , but its not updated here for some reason :)
please refer
https://github.com/sameerawickramasekara/wso2-test-grid/blob/b991889db24d00c3b76b19e53a8f219c17487945/core/src/main/java/org/wso2/testgrid/tinkerer/TinkererClient.java#L24

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java [hunk 14]
```diff
@@ -81,4 +84,50 @@ public String getInstanceName(String region, String instanceId) {
         }
         return instanceId;
     }
+
+    /**
+     *  This method will retrieve the instance username that is used to log into the instance via ssh.
+     *  E.g Ubuntu instance has username ubuntu
+     *  CentOS instance has username centos
+     *  The username must be present in the ami as a TAG with key USERNAME.
```
[the remark this answers is not in the document]
Comment (sameerawickramasekara) [reply deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java]: Added a doc entry to the quick start guide

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/api/DeploymentTinkerer.java (hunk 6 above)
Comment (kasunbg) [reply deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/api/DeploymentTinkerer.java]: .

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java [hunk 15]
```diff
@@ -49,6 +49,8 @@
     private static final Logger logger = LoggerFactory.getLogger(AWSProvider.class);
 
     private static Path configFilePath;
+    private static final String USERNAME_ENTRY = "USERNAME";
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java]: *OS_USERNAME_ENTRY

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java [hunk 16]
```diff
@@ -49,6 +49,8 @@
...
+    private static final String NAME_ENTRY = "Name";
```
Comment (kasunbg) [inline deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/providers/AWSProvider.java]: The purpose this constant is not clear.

File: deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java (hunk 9 above)
Comment (kasunbg) [reply deployment-tinkerer/src/main/java/org/wso2/testgrid/deployment/tinkerer/utils/SSHHelper.java]: .

**Pull request merged**